### PR TITLE
be more secure by validating both aud and iss claims

### PR DESCRIPTION
**Pull request by dsschult**
Be more secure by validating both aud and iss claims.

And also remove dup code from both Auth classes by making a base class.

### CODE REVIEW

File: rest_tools/utils/auth.py [hunk 1]
```diff
@@ -67,22 +110,13 @@ def validate(self, token, **kwargs):
...
-        except jwt.exceptions.InvalidAudienceError:
-            if 'audience' not in kwargs:
-                kwargs['audience'] = ['ANY']
-                ret = jwt.decode(token, self.pub_secret, issuer=self.issuer, algorithms=[self.algorithm], **kwargs)
-            else:
-                raise
-        if 'type' not in ret:
-            raise Exception('no type information')
-        return ret
+        return self._validate(token, self.pub_secret, issuer=self.issuer, required=['type'], **kwargs)
```
Comment (ric-evans) [inline rest_tools/utils/auth.py:113]: Is there any reason why you decided against calling (and creating) `super().validate(...)`?

File: rest_tools/utils/auth.py [hunk 2]
```diff
@@ -131,14 +165,6 @@ def validate(self, token, audience=None):
...
-            options = {}
-            kwargs = {}
-            if not audience:
-                options['verify_aud'] = False
-            elif audience.lower() == 'any':
-                kwargs['audience'] = ['ANY']
-            else:
-                kwargs['audience'] = [audience]
-            return jwt.decode(token, key, algorithms=['RS256','RS512'], options=options, **kwargs)
+            return self._validate(token, key, **kwargs)
```
Comment (ric-evans) [inline rest_tools/utils/auth.py:168]: See above

File: tests/utils/auth_test.py [hunk 3]
```diff
@@ -37,10 +37,42 @@ def test_auth_validate():
...
+
+    with pytest.raises(jwt.exceptions.InvalidIssuerError):
+        a._validate(tok, 'secret', issuers=['bar'])
+
+def test_auth_validate_iss_none():
+    a = auth.Auth('secret', issuer='foo')
+    now = time.time()
+    tok = a.create_token('subj', expiration=20, type='foo')
+    data = a._validate(tok, 'secret')
+    assert data['iss'] == 'foo'
```
Comment (ric-evans) [inline tests/utils/auth_test.py:74]: It's nice to see `jwt` has its own exception classes

**Review by ric-evans — CHANGES_REQUESTED** [review]

File: rest_tools/utils/auth.py (hunk 1 above)
Comment (dsschult) [reply rest_tools/utils/auth.py:113]: The main reason is that it's not meant to be used directly, so subclasses shouldn't be able to "pass" the method on to users.

**Review by ric-evans — APPROVED** [review]

**Pull request merged**